pkg/pgmodel/lreader/labels_reader.go

```diff
@@ -97,44 +93,49 @@ func (lr *labelsReader) LabelNames() ([]string, error) {
 	return labelNames, nil
 }
 
-// PrompbLabelsForIds returns protobuf representation of the label sets for
-// the provided label ids
-func (lr *labelsReader) PrompbLabelsForIds(ids []int64) (lls []prompb.Label, err error) {
-	ll, err := lr.LabelsForIds(ids)
-	if err != nil {
-		return
-	}
-	lls = make([]prompb.Label, len(ll))
-	for i := range ll {
-		lls[i] = prompb.Label{Name: ll[i].Name, Value: ll[i].Value}
+// LabelsForIdMap fills in the label.Label values in a map of label id => labels.Label.
+func (lr *labelsReader) LabelsForIdMap(idMap map[int64]labels.Label) error {
+	numIds := len(idMap)
+	ids := make([]interface{}, numIds) //type int64
+	lbs := make([]interface{}, numIds) //type labels.Label
+
+	//id=0 reserved for "no label for key" so this lookup would always fail. Should
+	//never have been passed in with the idMap to begin with.
+	_, present := idMap[0]
+	if present {
+		return fmt.Errorf("looking up a label for id 0")
 	}
-	return
-}
 
-// LabelsForIds returns label names and values for the supplied IDs.
-func (lr *labelsReader) LabelsForIds(ids []int64) (lls labels.Labels, err error) {
-	keys := make([]interface{}, len(ids))
-	values := make([]interface{}, len(ids))
-	for i := range ids {
-		keys[i] = ids[i]
+	i := 0
+	for id := range idMap {
+		ids[i] = id
+		i++
 	}
-	numHits := lr.labels.GetValues(keys, values)
+	numHits := lr.labels.GetValues(ids, lbs)
 
-	if numHits < len(ids) {
-		var numFetches int
-		numFetches, err = lr.fetchMissingLabels(keys[numHits:], ids[numHits:], values[numHits:])
+	if numHits < numIds {
+		var (
+			numFetches int
+			err        error
+		)
+
+		missingIds := make([]int64, numIds-numHits)
+		numFetches, err = lr.fetchMissingLabels(ids[numHits:], missingIds, lbs[numHits:])
 		if err != nil {
-			return
+			return err
+		}
+		if numFetches+numHits != numIds {
+			return fmt.Errorf("missing labels")
```

Comment:
should we also mention which all ids were missing the labels for better debugging?